README.md
# MQTT Hub

Turn your Homey into a HUB and let external applications discover & control all your connected devices.
  
## Introduction
With this app you can communicate with all your devices connected to Homey using MQTT.  
The MQTT Hub automatically broadcasts all your devices and capabilities. 
It will setup mqtt communication channels for external apps to discover and control the Homey devices.
Auto discovery protocols are implemented to simplify the setup and connection with external apps. 
Finally this app includes an `MQTT Device`, which can be used to add Virtual Devices with mappings between Homey capabilities and mqtt topics. 
  
## Functionality
The functionality of the MQTT Hub:
- Broadcast all available Homey devices and capabilities.
- Create communication channels for each device.
- Continously dispatch device state changes for all connected devices.
- Auto discovery of your Homey devices for external apps.
- Auto discovery of external devices using the Homie convention.
- Add the ability to remotely update the state of any device.
- Dispatch system info (memory, cpu, etc.) on a regular interval.
- Add Virtual MQTT Devices with mappings between Homey capabilities and mqtt topics.
  
All can be configured via extensive app settings and connection wizards.
  
## What can it be used for?
Some of the many possibilities:
- Integrate with other home automation systems: [OpenHab](https://www.openhab.org/), [Home Assistant](https://www.home-assistant.io/), [Domoticz](http://www.domoticz.com/), ...
- Create custom dashboards: [Lovelace UI](https://www.home-assistant.io/lovelace/), [TileBoard](https://community.home-assistant.io/t/tileboard-new-dashboard-for-homeassistant/57173), [HABPanel](https://www.openhab.org/docs/configuration/habpanel.html), [Node RED Dashboard](https://flows.nodered.org/node/node-red-dashboard), ...
- Create advanced flows and logic: [Node RED](https://nodered.org/), ...
- Use native mobile apps (3rd party): [MQTT Dash](https://play.google.com/store/apps/details?id=net.routix.mqttdash), ...
- Connect to virtual assistants (Google Home/Assistant): [gBridge](https://gBridge.io), ...
- Control external devices over MQTT by adding a virtual Homie Device
- ...
  
## MQTT?
[MQTT](http://mqtt.org/) is a lightweight communication protocol on top of TCP/IP and it's (becoming) the industry standard for IoT messaging (Internet of Things).
The [Homie Convention](https://homieiot.github.io/) (v3.0.1) is implemented to streamline the communication with external platforms by applying a communication standard.
 
#### MQTT client & broker
This app uses the [MQTT Client app](https://apps.athom.com/app/nl.scanno.mqtt) to communicate with a MQTT broker.
You can connect with any broker (e.g. [CloudMQTT](https://www.cloudmqtt.com/), [Mosquitto](https://mosquitto.org/) or [HiveMQ](https://www.hivemq.com/)). There is also a [MQTT broker app](https://apps.athom.com/app/nl.scanno.mqttbroker) available for Homey.
  
## Installation
1. Install an MQTT broker of your liking.
2. Install the MQTT Client app from the store and connect to your broker.
3. Install the MQTT Hub and let it discover & broadcast your devices.
4. Install any external application supporting MQTT and connect it to your broker.
5. Let your app discover Homeys devices (if it supports the Homie Convention). 
6. Configure your communication channels by observing & dispatching messages on the available topics.
7. Start communicating with homey.
  
TIP: [MQTT Explorer](https://mqtt-explorer.com/) is a nice tool to check the available communication channels.
  
## MQTT Device
An MQTT Device driver is available to control external MQTT devices from within Homey. This device can be added via the 'add Device' wizard. Select a custom device class and add functionality by adding capabilities mapped to MQTT topics. Incoming messages can be manipulated by adding a value template. Values can be grabbed from a JSON formatted message by using JSONPath or calculated using a mathJS expression. Outgoing messages can also be formatted using an output template. If JSON output is required, JSON-T style formatting is implemented. mathJS calculations are also allowed for outgoing data.  

## Communication
The hub allows several ways of communication.  
The default communication protocol is based on the [Homie Convention](https://homieiot.github.io/) (v3.0.1), for details see the [specification](https://homieiot.github.io/specification/).  
From their website: *The Homie convention defines a standardized way of how IoT devices and services announce themselves and their data on the MQTT broker*.
  
#### Auto discovery
External apps can automatically discover your devices connected to Homey by auto discovery. The Homie Convention supports this by design. Additionally [HA Discovery](https://www.home-assistant.io/docs/mqtt/discovery/) is implemented.  
  
#### MQTT Topic
The following message format is used for communication:  
`root`/`system name`/`device class (optional)`/`zone (optional)`/`device name`/`capability`/`command`  
  
NOTE: `root`, `system name`, `zone` & `device name` will all be normalized.  
i.e. All special characters will be removed, spaces are replaced by a dash (-) and all remaining text will be converted to lowercase characters.   
E.g. `Homey/light/Living room/Light tv/...` will become `homey/light/living-room/light-tv/...`
  
#### Customize
The MQTT Hub also allows a `custom` communication protocol with the ability to configure:
- topic structure (e.g. inject device class and/or zone)
- color format (HSV, RGB, channels)
- enable/disable dispatching of device states
- disable topic normalization
- dimensions & value scaling  
     
When using the command structure:
- Device id's will automatically be resolved from the device id, name or topic (in this order).
- The device name may contain either de original name or the normalized version.  
  
#### Birth & Last Will  
The MQTT Hub is able to broadcast a `birth` message on startup and a `last will` message on shutdown. This can be used by external applications to act on the Hub availability (`online`/`offline`).
  


## Change Log  
Can be found here: [CHANGE_LOG.md](/CHANGE_LOG.md).
  
## Final note ##
The repository is available at: https://github.com/harriedegroot/nl.hdg.mqtt  
If you want to contribute, just create a [pull-request](https://help.github.com/articles/about-pull-requests/) and I will take a look at it!

Do you like this app? Consider a donation to support development.
 
[![Donate][pp-donate-image]][pp-donate-link]

[pp-donate-link]: https://www.paypal.me/harriedegroot
[pp-donate-image]: https://img.shields.io/badge/Donate-PayPal-green.svg


CHANGE_LOG.md
## Change Log  

#### 3.6.1
Home Assistant auto discovery:
- group entities/capabilities per device
- allow optional device_id in discovery topic
- retain HomeAssistant config topics
- customize capability display names
- added default capabilities for sensor types: text, numeric & binary
- improved HA entity to capability matching

#### 3.6.0
- Home Assistant Discovery driver
- Renamed `setTopic` to `commandTopic`
- Fixed topic configuration for MQTT Devices created from discovery
- Improved discovery topic subscription
- Removed success popup when a new device is added from descovery

#### 3.5.5
- Resolved an issue for boolean handling in output templates

#### 3.5.4
MQTT Device:
- Selected icon styling improvements
- Prevent calling the set topic for state changes from mqtt
- Resolved a parsing error for string messages send to topics with an output template
- MathJS equality expression fix for comparing strings in value templates

#### 3.5.3
- MQTT Device: mimetype fix for uploading SVG icons

#### 3.5.2
- HA Discovery: Cover position template fix for HA version 2021.6+ ([issue 56](https://github.com/harriedegroot/nl.hdg.mqtt/issues/56))
- Security: Homey, jsonPath & lodash NPM package upgrades

#### 3.5.1
- Upload custom icon fixes
- Remove uploaded icons on cancel pairing OR uninstall of MQTT Device

#### 3.5.0
- MQTT Device icons

#### 3.4.2
- MQTTDevice: Display capability details
- Readme update

#### 3.4.1
- MQTTDevice: allow mathJS expressions as output template for outgoing messages
- MQTTDevice: removed default '{{value}}' output template

#### 3.4.0
- MQTTDevice: allow mathJS expressions as value template for incoming messages

#### 3.3.3
- Color temperature fix (HA => Homey)

#### 3.3.2
- Color temperature fix (Homey => HA)

#### 3.3.1
- MQTT Device: allow the same topic to be used for multiple capabilities

#### 3.3.0
MQTT Device additions:
- allow the same capability to be added multiple times
- provide a custom display name (capability title)
- edit capabilities while creating a new MQTT Device

#### 3.2.0
- Unsubscribe from topics
- Discovery: skip unknown devices

#### 3.1.9
- HA Dispatcher: REVERTED: removed device name from entity name
    
#### 3.1.8
- HA Dispatcher: removed device name from entity name
  
#### 3.1.7
- Skip non-homie device topics in homieDevice message handling
  
#### 3.1.6
- Logging: parse objects to JSON string for debugging puposes
  
#### 3.1.5
- lib bump: missing packages
  
#### 3.1.4
- Homey lib update to v2.9.0
  
#### 3.1.3
- Changed default HA status topic to: homeassistant/status
  
#### 3.1.2
- Added brand color
  
#### 3.1.1
- Fix for publishing '0' and 'false' payloads
  
#### 3.1.0
- MQTT device: grab input value from json messages (jsonPath)
- MQTT device: format output message payload to json (JSON-T template)
  
#### 3.0.3
- [FIX] MQTT device topic registration
  
#### 3.0.2
- HA discovery: set cover position fix (Fibaro shutters)
- Removed benefit message from settings page
  
#### 3.0.1
- FIX: changing settings of an MQTT Device
- FIX: MQTT Device onoff values (false, 0) + state updates
- Prevent message loops when the MQTT device state is updated by a message
- Keep the MQTT device topics in sync
- Add remove MQTT Device capabilities via advanced device config (json)
- Change MQTT device topics in advanced device config (json)
  
#### 3.0.0  
- Upgrade to Homey SDK v3
- Added flow card to trigger a broadcast
- Removed guid from default homie topic at app install
- Delayed the initial broadcast by 30 sec.
- Removed 'icon' from Home Assistant payloads
- cast enum values to string
  
#### 2.2.5  
- Added MQTT device setting to modify the `onoff` capability output (`true`/`false`, `on`/`off`, etc.)
- Cleanup MQTT Device topic settings
  
#### 2.2.4  
- Manage topics after the device is installed (device `advanced settings` menu)
- Added `percentage scaling` device setting
- Fixed a bug in the device manager when adding or removing Homey devices
- Increased max log lines to 1000 for debug level
  
#### 2.2.3  
- Flowcard Triggers & Actions
- Improved performance of Homie discovery devices list
  
#### 2.2.2  
- Updated Icons & Images
- Fixed a small styling issue in MQTT Device capabilities list
  
#### 2.2.1  
- Homie MQTT Discovery
  
#### 2.2.0  
- MQTT Device
  
#### 2.1.9  
- Retry topic subscriptions on failure & manual broadcast  
  
#### 2.1.8  
- HA Discovery: covers/blinds
    
#### 2.1.7  
- PR: MQTT Explorer url
- Updated readme
- Athom api update (2.1.178)
- Bug fixes & stability improvements

#### 2.1.6  
- [FIXED] Devices are not displayed in app settings on clean install
  
#### 2.1.5  
- Option to enable/disable all devices
- Performance & stability improvements 
  
#### 2.1.4  
- Display broadcast progress  
  
#### 2.1.3  
- Updated Homey api to version 2.1.172  
- Home Assistant Discovery: on_command_type for lights set to 'last'
    
#### 2.1.2  
- Homie Convention integer datatype fix
  
#### 2.1.1  
- Improved app startup & shutdown procedures 
- Resolved issues with Client & Hub app install order
- Resolved some issues in MQTT Client app (un)install handling
- Skip normalization of custom topics
- Introduced log levels for 20% speed improvement
- Home Assistant status topic fix
- Home Assistant binary sensors fix
- Fixed the detection of config changes
- General bug fixes & stability
  
#### 2.1.0  
- HA Discovery (beta)
- Simplified settings page + added many options
- App performance & stability (introduction of a MessageQueue)
- Birth & Last Will messages
- Topic clean-up (remove retained messages from the broker when disabling devices)
- Many BUG fixes
  
#### 2.0.6  
- Fixed a BUG in the Homie protocol regarding device state broadcasts  
  
#### 2.0.5  
- Topic fixes for system state dispatcher & command handler  
  
#### 2.0.4  
- Added instructions  
  
#### 2.0.3  
- Custom communication protocol
- Commands (rewrote some code to allow the command structure again)
- General bug fixes  
  

app.js
'use strict';

const DEBUG = process.env.DEBUG === '1';
if (DEBUG) {
    require('inspector').open(9229, '0.0.0.0', false);
}

const STARTUP_DELAY = 30 * 1000; // wait 30 sec. before starting the broadcasts

const Homey = require('homey');
const { HomeyAPI } = require('athom-api');
const MQTTClient = require('./mqtt/MQTTClient');
const MessageQueue = require('./mqtt/MessageQueue');
const Message = require('./mqtt/Message');
const TopicsRegistry = require('./mqtt/TopicsRegistry');
const normalize = require('./normalize');

// Services
const Log = require("./Log.js");
const DeviceManager = require("./DeviceManager.js");

// Dispatchers
const SystemStateDispatcher = require("./dispatchers/SystemStateDispatcher.js");
const HomieDispatcher = require("./dispatchers/HomieDispatcher.js");
const HomeAssistantDispatcher = require("./dispatchers/HomeAssistantDispatcher.js");

// Commands
const CommandHandler = require("./commands/CommandHandler.js");

// Birth & Last will
const BIRTH_TOPIC = '{deviceId}/hub/status';
const BIRTH_MESSAGE = 'online';
const WILL_TOPIC = '{deviceId}/hub/status';
const WILL_MESSAGE = 'offline';

const DEFAULT_LOG_LEVEL = 'info';

class MQTTHub extends Homey.App {

    async onInit() {
        try {
            Log.info('MQTT Hub is running...');

            this.homey.on('unload', () => this.uninstall());

            this.settings = this.homey.settings.get('settings') || {};
            this.birthWill = this.settings.birthWill !== false;

            Log.setLevel(DEBUG ? 'debug' : this.settings.loglevel || DEFAULT_LOG_LEVEL);
            Log.debug(this.settings, false, false);

            this.api = await HomeyAPI.forCurrentHomey(this.homey);

            try {
                this.system = await this._getSystemInfo();
            } catch (e) {
                Log.error('[boot] Failed to fetch system info');
                Log.error(e);
                this.system = {};
            }

            Log.debug("Update settings");
            this.initSettings();

            Log.debug("Initialize MQTT Client & Message queue");
            this.mqttClient = new MQTTClient(this.homey);
            this.messageQueue = new MessageQueue(this.mqttClient);
            this.topicsRegistry = new TopicsRegistry(this.messageQueue);

            // Suppress memory leak warning
            Log.debug("Suppress memory leak warning");
            this.api.devices.setMaxListeners(9999); // HACK

            // devices
            Log.debug("Initialize DeviceManager");
            this.deviceManager = new DeviceManager(this);

            Log.debug("Register DeviceManager");
            await this.deviceManager.register();

            // run
            Log.debug("Launch!");
            await this.start();

            // listen to broadcast flow card actions
            this.homey.flow.getActionCard('broadcast')
                .registerRunListener(async (args, state) => {
                    this.log('Broadcast triggered from flow');
                    try {
                        await this.refresh();
                    }
                    catch (error) {
                        this.log('Broadcast flow card trigger failed: ' + error.message);
                    }
                });

            this._initialized = true;
        }
        catch (e) {
            Log.error('[boot] Failed to initialize app');
            Log.error(e);
        }
    }
    

    initSettings() {
        let systemName = this.system.name || 'Homey';
        if (this.settings.deviceId === undefined || this.settings.systemName !== systemName || this.settings.topicRoot) {

            // Backwards compatibility
            if (this.settings.topicRoot && !this.settings.homieTopic) {
                this.settings.homieTopic = this.settings.topicRoot + '/' + (this.settings.deviceId || systemName);
                delete this.settings.topicRoot;
            }

            this.settings.systemName = systemName;
            if(!this.settings.deviceId) {
                let idx = systemName.lastIndexOf('-');
                this.settings.deviceId = idx === -1 ? systemName : systemName.substr(0, idx);
            }

            Log.debug("Settings initial deviceId: " + this.settings.deviceId);
            this.homey.settings.set('settings', this.settings);
            Log.debug("Settings updated");
        }
    }

    /**
     * Start Hub
     * */
    async start() {
        try {
            if (!this.mqttClient.isRegistered()) {
                Log.debug("Connect MQTT Client");
                await this.mqttClient.connect();
            }

            if (this.mqttClient.isRegistered()) {
                Log.info('start Hub');
                await this._sendBirthMessage();

                this.homey.setTimeout(async () => {
                    try {
                        await this.run();
                        Log.info('app running: true');
                    } catch(e) {
                        Log.error('[RUN] Hub initializasion failed');
                        Log.error(e);
                    }
                }, STARTUP_DELAY);
            } else {
                Log.debug("Waiting for MQTT Client...");
                this.mqttClient.onRegistered.subscribe(() => this.start(), true); // NOTE: Recursive
            }
        } catch (e) {
            Log.error('Failed to start Hub');
            Log.error(e);
        }
    }

    /**
     * Stop Hub
     * */
    async stop() {
        if (!this._running) return;
        this._running = false;

        Log.info('stop Hub');
        try {
            await this._sendLastWillMessage();
        } catch (e) {
            Log.error("Failed to send last will message on stop");
            Log.error(e);
        }
        try {
            Log.info("Disconnect MQTT Client");
            await this.mqttClient.disconnect();
        } catch (e) {
            Log.error("Failed to disconnect MQTTClient");
            Log.error(e);
        }

        this._stopCommunicationProtocol();
        await this._stopBroadcasters();
        this._stopCommands();
        delete this.protocol;

        this.messageQueue.stop();
        this.messageQueue.clear();

        Log.info('app running: false');
    }

    /**
     * Load configuration & run
     * Note: Called from start & settings changed
     * */
    async run(force) {
        if (force !== true && this._running) return;
        this._running = true;

        this._initProtocol();

        await this._startCommands();
        await this._startBroadcasters();
        await this._startHomeAssistantDiscovery();
        await this._startCommunicationProtocol();
    }

    _initProtocol() {
        this.protocol = this.settings.protocol || 'homie3';
        Log.info('Initialize communication protocol: ' + this.protocol);

        switch (this.protocol) {
            case "custom":
                this.settings.homieTopic = this.settings.customTopic;
                break;
            case "homie3":
            default:
                this.settings.homieTopic = normalize(this.settings.homieTopic);
                this.settings.topicIncludeClass = false;
                this.settings.topicIncludeZone = false;
                this.settings.normalize = true;
                this.settings.percentageScale = "int";
                this.settings.colorFormat = "hsv";
                this.settings.broadcastDevices = true;
                break;
        }

        // NOTE: All communication is based on the (configurable) Homie Convention...
        Log.info("Initialize HomieDispatcher");
        this.homieDispatcher = this.homieDispatcher || new HomieDispatcher(this);
        this.homieDispatcher.applySettings(this.settings);
    }

    async _startCommands() {
        if (this.settings.commands) {
            Log.info("start commands");
            // TODO: Refactor command handler with the abillity to register commands
            this.commandHandler = this.commandHandler || new CommandHandler(this);
            await this.commandHandler.init(this.settings); 
        } else {
            this._stopCommands();
        }
    }
    _stopCommands() {
        if (this.commandHandler) {
            Log.info("stop command handler");
            this.commandHandler.destroy();
            delete this.commandHandler;
        }
    }

    async _startBroadcasters() {
        if (this.settings.broadcastSystemState) {
            Log.info("start system state broadcaster");
            this.systemStateDispatcher = this.systemStateDispatcher || new SystemStateDispatcher(this);
            await this.systemStateDispatcher.init(this.settings);
        } else {
            this._stopBroadcasters();
        }
    }
    async _stopBroadcasters() {
        if (this.systemStateDispatcher) {
            Log.info("stop system state broadcaster");
            try {
                await this.systemStateDispatcher.destroy();
            } catch (e) {
                Log.error("Failed to destroy SystemState Dispatcher");
                Log.error(e);
            }
            delete this.systemStateDispatcher;
        }
    }
    
    async _startHomeAssistantDiscovery() {
        if (this.settings.hass) {
            Log.info("start Home Assistant Discovery");
            this.homeAssistantDispatcher = this.homeAssistantDispatcher || new HomeAssistantDispatcher(this);
            await this.homeAssistantDispatcher.init(this.settings, this.deviceChanges);
        } else {
            Log.info("stop Home Assistant Discovery");
            this._stopHomeAssistantDiscovery();
        }
    }
    _stopHomeAssistantDiscovery() {
        if (this.homeAssistantDispatcher) {
            Log.info("stop Home Assistant Discovery");
            this.homeAssistantDispatcher.destroy();
            delete this.homeAssistantDispatcher;
        }
    }

    async _startCommunicationProtocol() {
        // Register all devices & dispatch current state
        Log.info('Start communication protocol: ' + this.protocol);
        await this.homieDispatcher.init(this.settings, this.deviceChanges);
    }
    _stopCommunicationProtocol() {
        // NOTE: All communication is based on the (configurable) Homie Convention...
        if (this.homieDispatcher) {
            Log.info('stop communication protocol: ' + this.protocol);
            this.homieDispatcher.destroy();
            delete this.homieDispatcher;
        }
    }

    async _getSystemInfo() {
        Log.debug("get system info");
        const info = await this.api.system.getInfo();
        return {
            name: info.hostname,
            version: info.homey_version
        };
    }

    async getDevices() {
        if (this.deviceManager) {
            try {
                Log.debug("get devices");
                if (this.deviceManager && this.deviceManager.devices)
                    return this.deviceManager.devices;

                const api = await HomeyAPI.forCurrentHomey(this.homey);
                return await api.devices.getDevices();
            } catch (e) {
                Log.info("Failed to get Homey's devices");
                Log.error(e);
            }
        } else {
            return [];
        }
    }

    async getZones() {
        if (this.deviceManager) {
            try {
                Log.debug("get zones");
                if (this.deviceManager && this.deviceManager.zones)
                    return this.deviceManager.zones;

                const api = await HomeyAPI.forCurrentHomey(this.homey);
                return await api.zones.getZones();
            } catch (e) {
                Log.info("Failed to get Homey's zones");
                Log.error(e);
            }
        } else {
            return [];
        }
    }

    isRunning() {
        return this._running;
    }

    setRunning(running) {
        Log.info(running ? 'switch on' : 'switch off');
        if (this.mqttClient) {
            if (running) {
                this.start()
                    .then(() => Log.info("App running"))
                    .catch(error => Log.error(error));
            }
            else {
                this.stop()
                    .then(() => Log.info("App stopped"))
                    .catch(error => Log.error(error));
            }
        }
    }

    getState() {
        if (this.messageQueue) {
            const state = this.messageQueue.getState();
            //Log.debug(state);
            return state;
        }
        return {};
    }

    /**
     * Publish all device states
     * */
    async refresh() {
        Log.info('refresh');

        if (!this._initialized) return;

        if (this.mqttClient) {
            await this.mqttClient.retryFailedSubscriptions();
        }

        if (this.homeAssistantDispatcher) {
            this.homeAssistantDispatcher.dispatchState();
        }
        if (this.homieDispatcher) {
            this.homieDispatcher.dispatchState();
        }
    }

    async settingsChanged() {

        try {
            Log.info("Settings changed");
            this.settings = this.homey.settings.get('settings') || {};
            Log.debug(this.settings);

            // birth & last will
            if (this.settings.birthWill) {
                if (this.birthWill !== this.settings.birthWill) {
                    await this._sendBirthMessage();
                }
            } else {
                if (this.birthWill) {
                    await this._clearBirthWill();
                }
            }
            this.birthWill = this.settings.birthWill;

            // devices
            if (this.deviceManager) {
                this.deviceChanges = this.deviceManager.computeChanges(this.settings.devices);
                this.deviceManager.setEnabledDevices(this.settings.devices);
            }

            if (this._initialized) {
                await this.run(true);
            }

            // clean-up all messages for disabled devices
            for (let deviceId of this.deviceChanges.disabled) {
                if (typeof deviceId === 'string') {
                    this.topicsRegistry.remove(deviceId, true);
                }
            }

            // clean-up
            delete this.deviceChanges; 

        } catch (e) {
            Log.error("Failed to update settings");
            Log.error(e);
        }
    }

    get _birthTopic() {
        const deviceId = this.settings.normalize !== false ? normalize(this.settings.deviceId) : this.settings.deviceId;
        return (this.settings.birthTopic || BIRTH_TOPIC).replace('{deviceId}', deviceId);
    }
    get _willTopic() {
        const deviceId = this.settings.normalize !== false ? normalize(this.settings.deviceId) : this.settings.deviceId;
        return (this.settings.willTopic || WILL_TOPIC).replace('{deviceId}', deviceId);
    }
    async _sendBirthMessage() {
        Log.debug("Send birth message");
        if (this.mqttClient && this.settings.birthWill !== false) {
            const msg = this.settings.birthMessage || BIRTH_MESSAGE;
            return await this.mqttClient.publish(new Message(this._birthTopic, msg, 1, true));
        }
    }
    async _sendLastWillMessageAndReleaseAllTopics() {
        Log.debug("Send last will message");
        if (this.mqttClient && this.settings.birthWill !== false) {
            const msg = this.settings.willMessage || WILL_MESSAGE;
            const result = await this.mqttClient.publish(new Message(this._willTopic, msg, 1, true));
            await this.mqttClient.release(); // Notify client to release all topics used by the app
            return result;
        }
    }
    async _clearBirthWill() {
        await this.mqttClient.publish(new Message(this._birthTopic, null, 1, true));
        await this.mqttClient.publish(new Message(this._willTopic, null, 1, true));
    }

    uninstall() {
        try {
            this._sendLastWillMessageAndReleaseAllTopics()
                .then(() => this.mqttClient.disconnect().catch(e => Log.error(e)))
                .catch(error => {
                    Log.error("Failed to send last will message at uninstall");
                    Log.error(error);
                    this.mqttClient.disconnect().catch(e => Log.error(e));
                });
        } catch(e) {
            // nothing...
        }
    }
}

module.exports = MQTTHub;